lib/vast/Dialect/HighLevel/HighLevelTypes.cpp

```diff
@@ -29,19 +30,70 @@ namespace vast::hl
         VAST_UNIMPLEMENTED;
     }
 
-    mlir::FunctionType getFunctionType(Type type)
-    {
+    Type getBottomTypedefType(TypedefType def, Module mod) {
+        auto type = getTypedefType(def, mod);
+        if (auto ty = type.dyn_cast< TypedefType >()) {
+            return getBottomTypedefType(ty, mod);
+        }
+        return type;
+    }
+
+    Type getTypedefType(TypedefType type, Module mod) {
+        auto name = type.getName();
+        for (const auto &op : mod) {
+            if (auto def = mlir::dyn_cast< TypeDefOp >(&op)) {
+                if (def.getName() == name) {
+                    return def.getType();
+                }
+            }
+        }
+
+        VAST_UNREACHABLE("unknown typedef name");
+    }
+
+    mlir::FunctionType getFunctionType(Type type, Module mod) {
         if (auto ty = type.dyn_cast< mlir::FunctionType >())
             return ty;
         if (auto ty = type.dyn_cast< LValueType >())
-            return getFunctionType(ty.getElementType());
+            return getFunctionType(ty.getElementType(), mod);
         if (auto ty = type.dyn_cast< ParenType >())
-            return getFunctionType(ty.getElementType());
+            return getFunctionType(ty.getElementType(), mod);
         if (auto ty = type.dyn_cast< PointerType >())
-            return getFunctionType(ty.getElementType());
+            return getFunctionType(ty.getElementType(), mod);
+        if (auto ty = type.dyn_cast< TypedefType >()) {
+            return getFunctionType(getTypedefType(ty, mod), mod);
+        }
+
+        type.dump();
```

Comment:
```suggestion
```

Forgotten debug print? Or is that supposed to be part of report if `UNREACHABLE` is reached?